components/controller/src/identifier_controller.rs

```diff
@@ -673,45 +674,76 @@ impl IdentifierController {
             .into_iter()
             .flatten()
             .collect::<Vec<_>>();
-        let n = receipts.len();
+        let mut n = 0;
 
-        // TODO: don't send the same receipt twice.
         for rct in receipts {
-            for wit in wits {
-                // TODO: don't send receipt to witness who created it.
+            let rct_digest = rct.body.event.receipted_event_digest.clone();
+            let rct_wit_ids = self.get_wit_ids_of_rct(&rct)?;
+
+            // Don't send the same receipt twice.
+            if rct_wit_ids.iter().all(|rct_wit_id| {
+                self.broadcasted_rcts
+                    .contains(&(rct_digest.clone(), rct_wit_id.clone()))
+            }) {
+                continue;
+            }
+
+            for dest_wit_id in dest_wit_ids {
+                // Don't send receipt to witness who created it.
+                // TODO: this only works if the target witness ID is a BasicPrefix.
+                if let IdentifierPrefix::Basic(dest_wit_id) = dest_wit_id {
+                    if rct_wit_ids.contains(dest_wit_id) {
+                        continue;
+                    }
+                }
+
+                // TODO: what if only some witnesses are online?
                 self.source
                     .send_message_to(
-                        wit,
+                        dest_wit_id,
                         Scheme::Http,
                         Message::Notice(Notice::NontransferableRct(rct.clone())),
                     )
                     .await?;
+
+                n += 1;
             }
 
             // Remember event digest and witness ID to avoid sending the same receipt twice.
-            let digest = rct.body.event.receipted_event_digest;
-            for sig in rct.signatures {
-                match sig {
-                    Nontransferable::Indexed(sigs) => {
-                        for sig in sigs {
-                            let wits = self.source.storage.get_witnesses_at_event(
-                                rct.body.event.sn,
-                                &self.id,
-                                &digest,
-                            )?;
-                            self.broadcasted_rcts
-                                .insert((digest.clone(), wits[sig.index as usize].clone()));
-                        }
+            for rct_wit_id in rct_wit_ids {
+                self.broadcasted_rcts
+                    .insert((rct_digest.clone(), rct_wit_id));
```

Comment:
How about inserting only witnesses for who sending receipts was successful? Then in the next `broadcast_receipts` call we'll try again for those, who didin't get it.